Rewrite, and feature additions

tesselslate opened the pull request:
Rewrote most of the entire map editor.

Notable reorganization:
- Map parsing has been moved to the `Map` namespace instead of `MapStructure`.
- The `Map` namespace also only contains map parsing now, rather than bundling editing and rendering as well.
- The `Map` class no longer exists, it is now `Level`. In turn, what was previously the `Level` class is now the `Room` class.
- Level rendering has been moved to the `Editor.Render` namespace.
- CPU usage really isn't the main bottleneck with rendering entities, etc. StaticTexture is probably unreasonable to use for those, but it is still very useful for tilemaps. It is also in the `Editor.Render` namespace.
- Instead of using raw `StaticTexture` arrays, there is now a `TextureMap` class which hopefully implements such functionality more cleanly.
- TileGrids now store their data with shorts, rather than integers. Probably not a large memory saving, but still an easy change that should reduce memory usage somewhat.
- The `GFX` class is now in `Mod.Content`.
- Startup takes place mainly in the `Core.Boot` namespace, which supports running startup tasks on multiple threads. However, loading textures on multiple threads appears to be problematic sometimes. It's instead done after all tasks have completed.
- Entities **can not use predefined static DrawableTextures.** This is because plugin loading takes place *before* gameplay textures are loaded. Perhaps adding a more robust plugin system with "module" classes to perform post-asset-load operations would be useful in the future.

Feature additions:
- (More of a bug fix) Replaced the prebuilt nfd_d.dll with a release build. 5x smaller and also actually works on other people's systems now.
- (More of a bug fix) Using keybinds which open nativefiledialog should no longer cause FNA to consider certain keys "stuck." This is done by hiding the window and immediately displaying it again. This is a janky workaround, but I can't find a better way to do it. Perhaps expanding the Shortcut system to allow for triggering the shortcut *after* keys are released would be helpful. The shortcut system will likely need more work anyway as more editor keybinds are implemented.
- Partial ImGui backend implementation. The desktop cursor is now changed based on the ImGui state. Additionally, a clipboard getter (for pasting) has been implemented (hopefully with cross-platform compatibility) with SDL. I attempted to implement a setter for actually *copying* text, but doing so would cause ImGui to crash regardless of what I did for some reason. Windows users will still be able to copy text. Users on other platforms (in the future, hopefully) likely will not be able to. It's a map editor anyway, who needs to copy text from it?
- Room addition/configuration/deletion - This is implemented through the EditorAction system. You can add new rooms, delete rooms, and undo these changes (and it appears to work.) Hopefully, this doesn't cause absurd memory usage when the editor is open as lots of rooms are being deleted. Perhaps it would be a good idea to add some sort of history cap in the future, where only a certain amount of actions are stored (perhaps 1k, or 5k)
- View menu - toggle the visibility of certain objects in the map.
- Trigger rendering - renders triggers. Not much else to say.
- Placement system - entities are no longer present in the placement menu by default. You now have to explicitly register placements.
- Slightly less bad GUI system. `Window` classes can be instantiated and registered through `Engine`, which will render them and automatically remove them when they are closed. Less hard-coded stuff.
- Object tiles
  - Object tiles are implemented somewhat weirdly. "Fake" tiles (Dash Block, Fake Wall, etc) can make use of the Autotiler to generate a "fake texture map" for them. If object tiles are visible, it will attempt to determine which object tiles will collide with the "fake tile" and make those colliding textures invisible. The object tiles will render normally where they should.
  - This works with switching the Entity/Object Tiles visibility since switching layer visibility in the View menu results in the entire map being rerendered.

tesselslate committed:
Rewrite map parsing

tesselslate committed:
Improvements and additions

- Added better boot/launch system, with multithreading and a loading animation.
- Added better settings system with less hardcoded stuff.
- Various changes.

tesselslate committed:
Basic map rendering

Reimplemented autotiler, as well as a separate map renderer. Moved texture loading to its own task performed on the main thread after other boot tasks, as multithreaded GPU use appears to potentially cause some issues.

tesselslate committed:
Rudimentary objtiles support

tesselslate committed:
Decal/trigger/filler rendering

tesselslate committed:
Add entity system

tesselslate committed:
Basic entities

tesselslate committed:
UI additions

tesselslate committed:
Tools, part 1

tesselslate committed:
Tile tools

tesselslate committed:
Entity tool, part 1

tesselslate committed:
Entity tool finished

tesselslate committed:
Reimplement entities

tesselslate committed:
Update nfd_d.dll

tesselslate committed:
Room actions

tesselslate committed:
Fix tools updating when switching them

tesselslate committed:
Cache sprites with Lazy

Suggested by Ja. Replaces the previous `static` DrawableTextures, since they don't work now.

AntonReinhard commented on Starforge/Core/Logger.cs:
```diff
@@ -6,53 +6,96 @@ namespace Starforge.Core {
+        /// <param name="msg">The message to write.</param>
         public static void Log(string msg) {
-            if (Level >= LogLevel.Info) {
+            if (Active &&Level >= LogLevel.Info) {
```
missing space smh

apppppppple commented on Starforge.Vanilla/Entities/IntroCar.cs:
```diff
@@ -1,20 +1,25 @@
-        private static readonly DrawableTexture bodySprite = GFX.Gameplay["scenery/car/body"];
-        private static readonly DrawableTexture wheelsSprite = GFX.Gameplay["scenery/car/wheels"];
+        private static Lazy<DrawableTexture> bodySprite = new Lazy<DrawableTexture>(() => GFX.Gameplay["scenery/car/body"];
+        private static Lazy<DrawableTexture> wheelsSprite = new Lazy<DrawableTexture>(() => GFX.Gameplay["scenery/car/wheels"];
```
Missing parenthesis?

apppppppple commented on Starforge/Mod/EntityRegistry.cs:
```diff
@@ -0,0 +1,92 @@
+            return new List<string>(EntityCreators.Keys);
+        }
+
+        /// <returns>A list of all registered entity placements.</retrurns>
```
Misspelled returns element

apppppppple commented on Starforge/Core/Boot/StartupHelper.cs:
```diff
@@ -0,0 +1,182 @@
+        /// Checks if the given folder appears to be a valid Celeste install.
+        /// </summary>
+        /// <param name="location">The path to check.</param>
+        /// <returns></returns>
```
Remove empty returns element

tesselslate committed:
Miscellaneous cleanup

apppppppple commented on Starforge/Core/Input.cs:
```diff
@@ -0,0 +1,117 @@
+        /// Determines if any mouse button states have changed.
+        /// </summary>
+        /// <returns></returns>
+        public bool HasAny() => 
```
Remove empty returns element

apppppppple commented on Starforge/Core/Input.cs:
```diff
@@ -0,0 +1,117 @@
+        /// <summary>
+        /// Returns an array of all currently pressed keys.
+        /// </summary>
+        /// <returns></returns>
```
Remove empty returns element

tesselslate committed:
Fix empty xmldoc comments

AntonReinhard commented on Starforge/Editor/Autotiler.cs:
```diff
@@ -0,0 +1,261 @@
+        /// <param name="xmlPath">The path to the file containing the tileset definitions.</param>
+        public Autotiler(string xmlPath) {
+            // Create tile randomization array
+            TileRand = new byte[1048576];
```
This seems to be pretty overkill, since you only ever use [x+yInc] in this array, which will probably not be this big. So maybe just make this a lot smaller, like 512 or something (also make that a constant please xD) and then when you get a value from it just do [(x + yInc) % size]

AntonReinhard commented on Starforge/Editor/Camera.cs:
```diff
@@ -91,24 +113,27 @@ public class Camera {
-            Viewport = Engine.Instance.GraphicsDevice.Viewport;
-            Update();
-        }
+        /// <summary>
```
Why not make one function that has a zoom multiplier? Or at least add a third for that

tesselslate replied:
What do you mean by this exactly?

AntonReinhard replied:
Instead of just being able to zoom by *2 or /2, make the zoom multiplier an argument. You could for example make the zoom "speed" make a setting when doing that. Also the code is kinda redundant except for that one /2 or *2 line

tesselslate replied:
As for the first part - zoom is always halved or doubled because going to other values between results in individual pixels of the map not scaling properly and turning into rectangles instead of perfect squares, which looks bad.

The code could probably be improved as well, yes, but I don't remember getting it to center on the mouse cursor properly without.. what is already there. The Update() calls at the end can be removed, though.

AntonReinhard commented on Starforge/Editor/Camera.cs:
```diff
@@ -47,24 +53,40 @@ public class Camera {
         }
 
+        /// <summary>
+        /// Moves the camera the specified amount.
```
I'm assuming positive x value moves to right and positive y value moves down? Maybe add directions to the comment

AntonReinhard commented on Starforge/Editor/Camera.cs:
```diff
@@ -91,24 +113,27 @@ public class Camera {
+        /// <param name="pos">The position to center the camera on when zooming.</param>
         public void ZoomOut(Vector2 pos) {
             GotoCentered(ScreenToReal(pos));
 
```
This doesn't seem to do any actual zooming?

tesselslate replied:
`Zoom /= 2f;`

AntonReinhard replied:
Oh, it didn't show those lines cause they didn't change and i didn't notice nvm

AntonReinhard commented on Starforge/Editor/EditorState.cs:
```diff
@@ -0,0 +1,158 @@
+        /// Undoes the last action, if one is available.
+        /// </summary>
+        public void Undo() {
+            if (PastActions.Count > 0) {
```
You could use your CanUndo() Property for this, no?

AntonReinhard commented on Starforge/Editor/EditorState.cs:
```diff
@@ -0,0 +1,158 @@
+        /// Applies the previously undone action, if one is available.
+        /// </summary>
+        public void Redo() {
+            if (FutureActions.Count > 0) {
```
Same as Undo

AntonReinhard commented on Starforge/Editor/EditorState.cs:
```diff
@@ -0,0 +1,158 @@
+    /// <summary>
+    /// Contains information about the current state of the editor.
+    /// </summary>
+    public class EditorState {
```
Is EditorState the right name for this? The name doesn't sound to me like you can add/remove rooms and such on it

tesselslate replied:
It contains general info about the editor (loaded level, etc) as well.

AntonReinhard commented on Starforge/Editor/Render/LevelRender.cs:
```diff
@@ -0,0 +1,300 @@
+            room.FGDecals = CreateDecalTextureList(room.Room.ForegroundDecals);
+
+            // Generate tiles
+            if (room.TilesDirty) {
```
I don't really like how the room's TilesDirty is checked here, but the general room.Dirty is checked outside of the function somewhere else. Shouldn't the room take care of its own rendering anyways? And especially its own dirty flags

tesselslate replied:
I personally think it makes more sense to leave the Dirty check outside of the RenderRoom function, although there is room for debate on this. The function is used in 4 places:
- `LevelRender.Render` (x2)
- `MapEditor.LoadLevel` (prerenders level)
- `MapEditor.RerenderAll` (rerenders all rooms)

I think it makes more sense to leave the check outside of the `RenderRoom` function itself because then it makes it more clear that `LevelRender.Render` is only rerendering the rooms if they need to be rerendered, not calling the function regardless and then having it check there.

AntonReinhard commented on Starforge/Editor/Render/LevelRender.cs:
```diff
@@ -0,0 +1,300 @@
+    }
+
+    public enum RenderFlags {
+        None = 0,
```
Could use 0, 1 << 0, 1 << 1, 1 << 2 and so on instead of 0, 1, 2, 4 etc. More expressive imo and you can't accidentally mistype a number

AntonReinhard commented on Starforge/Editor/Render/StaticTexture.cs:
```diff
@@ -60,8 +61,12 @@ public struct StaticTexture {
         }
     }
-}
+}
```
Can we please add a new line at the end of every file again? :)

AntonReinhard commented on Starforge/Editor/Render/TextureMap.cs:
```diff
@@ -0,0 +1,62 @@
+        /// <returns>The texture at the given position.</returns>
+        public StaticTexture this[int x, int y] {
+            get {
+                if (x < 0 || x > Width - 1 || y < 0 || y > Height - 1) return default;
```
I like to use `>= something` instead of `> something - 1`, but just preference probably. Just used to it from for loops

AntonReinhard commented on Starforge/Editor/Tools/TileRectangleTool.cs:
```diff
@@ -1,92 +1,75 @@
 namespace Starforge.Editor.Tools {
-
     public class TileRectangleTool : Tool {
+        private Rectangle Hint = new Rectangle(-8, -8, 8, 8);
```
Why here and in the other tools -8, -8 now as x and y of the Hints? It doesn't seem obvious to me so maybe add a comment

tesselslate replied:
Tool locations are set OoB because they don't update their position when you first select them, causing them to just appear and stay in the corner until you move your mouse from the tool menu. Will add a comment

AntonReinhard commented on Starforge/Editor/UI/Menubar.cs:
```diff
@@ -0,0 +1,210 @@
+            }
+
+            if (ImGui.BeginMenu("Edit")) {
+                if (ImGui.MenuItem("Undo", "CTRL+Z", false, Engine.MapLoaded && MapEditor.Instance.State.CanUndo())) MapEditor.Instance.State.Undo();
```
Undo() already checks if it can undo itself, this check is redundant

tesselslate replied:
CanUndo is still checked here so you can see if you have any changes which can be undone. If CanUndo is false, the menu item is then greyed out so you can see that there are no changes you can undo.

AntonReinhard commented on Starforge/Map/Style.cs:
```diff
@@ -0,0 +1,70 @@
+                if (style.Name == "parallax") {
+                    // Apply styleground
+                    return new Parallax(style, root);
+                } else {
```
For early error catching and possibly extensibility(? idk) it might be worth to still check the Name in this case too, and if it's neither "parallax" nor "effect" then log it or something. Unless this works differently, then ignore me

tesselslate replied:
There is no `effect`. The names of effect styles are simply the ID of the effect (e.g. `windsnow`, `snowfg`, etc). It's a bit weird, but it's how it is.

AntonReinhard commented on Starforge/Mod/EntityRegistry.cs:
```diff
@@ -0,0 +1,92 @@
+                if (!type.IsSubclassOf(typeof(Entity))) return;
+
+                EntityDefinitionAttribute attr = type.GetCustomAttribute<EntityDefinitionAttribute>();
+                if (attr != null) {
```
Inverting this and the next two checks of this kind and returning immediately can save a lot of indentation

AntonReinhard commented on Starforge/Util/MiscHelper.cs:
```diff
@@ -72,35 +84,8 @@ public static class MiscHelper {
-
-            return array;
+        public static int RandInt(int seed, int max) {
+            return RandBytes[seed] % max;
```
seed % 2^16 to be safe

AntonReinhard replied:
actually just make the parameter an unsigned short

AntonReinhard reviewed (approved):
pogeline i'm through

tesselslate committed:
Fixes and improvements

Fixed feedback from Ruby's review comments. Cleaned up solution (thanks jade for pointing that out) and added editorconfig file to add final new lines when saving files.

Rewrote some of MapPacker for.. some performance benefit, I think? In testing with Any CPU, x86, and various versions of .NET Framework and Core, it seemed to be faster sometimes, and other times it would fail spectacularly. Either way, it isn't noticeably slower. So I'll just assume it's faster (:

AntonReinhard commented on Starforge/Util/MiscHelper.cs:
```diff
@@ -85,7 +86,7 @@ public static class MiscHelper {
 
         public static int RandInt(int seed, int max) {
-            return RandBytes[seed] % max;
+            return RandBytes[seed] % (max % ushort.MaxValue);
```
I think you mixed something up here, you wanna do
```suggestion
            return RandBytes[seed % ushort.MaxValue] % max;
```
Since RandBytes only contains bytes, i.e. values from 0 to 255 that `% (max % ushort.MaxValue)` would only have questionable effects in rare cases.

tesselslate committed:
Fix RandBytes

Co-authored-by: Anton Reinhard <s1509337@msx.tu-dresden.de>

WoofWoofDoggo commented:
@AntonReinhard @JaThePlayer Are you both fine with this being merged into main?
